sklearn/tree/tests/test_monotonic_tree.py

```diff
@@ -0,0 +1,348 @@
+    test = np.arange(n_samples_train, n_samples)
+    X_train = X[train]
+    y_train = y[train]
+    X_test = np.copy(X[test])
+
+    X_test_incr = np.copy(X_test)
+    X_test_decr = np.copy(X_test)
+    X_test_incr[:, 0] += 10
+    X_test_decr[:, 1] += 10
+    monotonic_cst = np.zeros(X.shape[1])
+    monotonic_cst[0] = 1
+    monotonic_cst[1] = -1
+
+    classifiers = CLF_TREES.copy()
+    classifiers.update({"GradientBoostingClassifier": GradientBoostingClassifier})
+    for name, TreeClassifier in classifiers.items():
+        if depth_first:
+            est = TreeClassifier(max_depth=None, monotonic_cst=monotonic_cst)
+        else:
+            est = TreeClassifier(
+                max_depth=None,
+                monotonic_cst=monotonic_cst,
+                max_leaf_nodes=n_samples_train,
+            )
+        if hasattr(est, "random_state"):
+            est.set_params(**{"random_state": global_random_seed})
+        if hasattr(est, "n_estimators"):
+            est.set_params(**{"n_estimators": 5})
+        est.fit(X_train, y_train)
+        y = est.predict_proba(X_test)[:, 1]
+
+        # increasing constraint
+        y_incr = est.predict_proba(X_test_incr)[:, 1]
+        # y_incr should always be greater than y
+        assert np.all(y_incr >= y)
+
+        # decreasing constraint
+        y_decr = est.predict_proba(X_test_decr)[:, 1]
+        # y_decr should always be lower than y
+        assert np.all(y_decr <= y)
+
+
+@pytest.mark.parametrize("depth_first", (True, False))
+def test_montonic_constraints_regressions(depth_first, global_random_seed):
+    n_samples = 1000
+    n_samples_train = 900
+    # Build a classification task using 3 informative features
+    X, y = make_regression(
+        n_samples=n_samples, n_features=5, n_informative=5, random_state=0
+    )
+    train = np.arange(n_samples_train)
+    test = np.arange(n_samples_train, n_samples)
+    X_train = X[train]
+    y_train = y[train]
+    X_test = np.copy(X[test])
+    X_test_incr = np.copy(X_test)
+    X_test_decr = np.copy(X_test)
+    X_test_incr[:, 0] += 10
+    X_test_decr[:, 1] += 10
+    monotonic_cst = np.zeros(X.shape[1])
+    monotonic_cst[0] = 1
+    monotonic_cst[1] = -1
+    regressors = REG_TREES.copy()
+    regressors.update({"GradientBoostingRegressor": GradientBoostingRegressor})
+
+    for name, TreeRegressor in regressors.items():
+        if depth_first:
+            est = TreeRegressor(max_depth=None, monotonic_cst=monotonic_cst)
+        else:
+            est = TreeRegressor(
+                max_depth=None,
+                monotonic_cst=monotonic_cst,
+                max_leaf_nodes=n_samples_train,
+            )
+        if hasattr(est, "random_state"):
+            est.set_params(random_state=global_random_seed)
+        if hasattr(est, "n_estimators"):
+            est.set_params(**{"n_estimators": 5})
+        est.fit(X_train, y_train)
+        y = est.predict(X_test)
+        # increasing constraint
```

Comment:
I would be rather strict on the wording.

```suggestion
        # Monotonically increase constraint applies on the positive class
```

This rewording also applies in others places for both monotonically increase constraint and monotonically decrease constraint.

Reply:
If "increase" is a noun as in "the increase", we should not use an adverb like "monotonically" but "Monotonic increase".

However, when it's followed but "constraints", both "Monotonically increasing constraint" and "Monotonic increase constraint" feel incorrect or ambiguous to me.

Reply:
Maybe "Monotonic increase constraint" is good in retrospect.

```suggestion
        # Monotonic increase constraint applies on the positive class
```

Reply:
To be exhaustive and clear, the following terms seems the most appropriate to me:
 - "monotonicity constraints" to refer to all of them (encoded by -1, 0, 1) which would replace "monotonic constraints"
 - "monotonic increase constraints" (encoded by 1) which would replace "positive constraints"
 - monotonic decrease constraints (encoded by -1) which would replace "negative constraints"
 
What do you think?

Reply:
I agree.

We can also keep the phrasing "monotically increasing" in comments where we speak about the behavior of `y_pred` given a specific input feature without referring to the constraint itself.

Reply:
Resolved by b9c7a90838bc84fcc9a6939604ddb51bfc52d316.